install/universal_pkg/setup.py

```diff
@@ -0,0 +1,226 @@
+import ctypes
+import pkg_resources
+import sys
+from typing import List, Optional
+
+from setuptools import setup
+
+
+VERSION = '9.5.0'
+
+# List of packages supported by this version of CuPy.
+PACKAGES = [
+    'cupy-cuda102',
+    'cupy-cuda110',
+    'cupy-cuda111',
+    'cupy-cuda112',
+    'cupy-cuda113',
+    'cupy-cuda114',
+    'cupy-cuda115',
+    'cupy-rocm-4-0',
+    'cupy-rocm-4-1',
+    'cupy-rocm-4-2',
+    'cupy-rocm-4-4',
+]
+
+# List of packages NOT supported by this version of CuPy.
+PACKAGES_OUTDATED = [
+    'cupy-cuda80',
+    'cupy-cuda90',
+    'cupy-cuda91',
+    'cupy-cuda92',
+    'cupy-cuda100',
+    'cupy-cuda101',
+]
+
+# List of sdist packages.
+PACKAGES_SDIST = [
+    'cupy',
+]
+
+
+class AutoDetectionFailed(Exception):
+    pass
+
+
+def _log(msg: str) -> None:
+    sys.stdout.write(f'[cupy-wheel] {msg}\n')
+    sys.stdout.flush()
+
+
+def _get_version_from_library(
+        libnames: List[str], funcname: str) -> Optional[int]:
+    """Returns the library version from list of candidate libraries."""
+
+    for libname in libnames:
+        try:
+            _log(f'Looking for library: {libname}')
+            runtime_so = ctypes.CDLL(libname)
+            break
+        except Exception as e:
+            _log(f'Failed to open {libname}: {e}')
+            return None
+    else:
+        _log('No more candidate library to find')
+        return None
+
+    func = getattr(runtime_so, funcname, None)
+    if func is None:
+        raise AutoDetectionFailed(
+            f'{libname}: {func} could not be found')
+    func.restype = ctypes.c_int  # cudaError_t
+    func.argtypes = [ctypes.POINTER(ctypes.c_int)]
+    version_ptr = ctypes.c_int()
+    retval = func(version_ptr)
+    if retval != 0:
+        raise AutoDetectionFailed(
+            f'{libname}: {func} returned error: {retval}')
+    version = version_ptr.value
+    _log(f'Detected version: {version}')
+    return version
+
+
+def _get_cuda_version() -> Optional[int]:
+    """Returns the detected CUDA version or None."""
+
+    if sys.platform == 'linux':
+        libnames = ['libcudart.so']
+    elif sys.platform == 'win32':
+        libnames = ['cudart64.dll']
+    else:
+        _log(f'CUDA detection unsupported on platform: {sys.platform}')
+        return None
+    _log(f'Trying to detect CUDA version from candidates: {libnames}')
+    version = _get_version_from_library(libnames, 'cudaRuntimeGetVersion')
```

Comment:
A subtlety here: This relies on the installing machine to have a GPU. For a reason I can't recall now, we chose to look up `CUDA_VERSION` instead, but with hindsight it created more problems that it solved, so this is better 👍 (Also as a reminder to self that I need to fix it...)
https://github.com/NVIDIA/cuQuantum/blob/8bd2f77bb6a03ba49b69e8706c9d5fd94752b95c/python/setup.py#L107-L108